Add Altair validator guide and initial p2p spec (#2201)
ethereum/eth2.0-specs: specs/lightclient/validator.md

## Prerequisites

All terminology, constants, functions, and protocol mechanics defined in the [HF 1 -- The Beacon Chain](./beacon-chain.md) doc are requisite for this document and used throughout. Please see this document before continuing and use as a reference throughout.

## Custom types

| Name | SSZ equivalent | Description |
| - | - | - |
| `SyncCommitteeIndex` | `uint64` | Index of a validator within a sync committee |

## Containers

### SyncCommitteeContribution

```python
class SyncCommitteeContribution(Container):
    # Slot during which this contribution is broadcast
    slot: Slot
    # Position of the validator within the sync committee

>     index: SyncCommitteeIndex

Reviewer: Pros of defining new type -- `SyncCommitteeIndex` -- instead of just using a `ValidatorIndex`. Presumable if you can verify this signature, you have map of validators <-> sync_committee_shuffling so both should be approximately equivalent

With `SyncCommitteIndex`, you do a lookup with the shuffling to get the validator index and verify the signature
With `ValidatorIndex`, you can verify signature direct, and need to make sure the validator index is contained within the current shuffling

Reviewer: Ah, I see. The idea here is that if a validator is duplicated in the shuffling they would send two messages.

I'm not sure it's worth the new type here and duplicate message on network. The client can resolve a duplicate assignment by including the signature twice in the aggregate.

That said, I see *why* this is here now. open to discuss more

>     # Signature by the validator over the block root of the _previous_ slot

Reviewer 2: Are we using the _previous_ slot opposed to the _current_ slot for better stability?

Author: that falls out as a side effect but i'd say the bigger reason is just that we use blockspace to include the sync committee's proof about a given block -- if we signed over the block root of the current slot and then included that data into the current block, it would change the current block root so we are back to square one.

Reviewer 2: Ah i see. I thought it would have worked like attestation

slot N at 0: submit block (package attestations from N-1)
slot N at 1/3: submit attestations and sync committee objects

slot N+1 at 0:  submit block (package attestations and sync committee object from N)

So sync_committee object for current slot is ok because it's getting included in the next slot, this is similar to attestation handling today

Reviewer: Ah, I think y'all are saying the same thing.

If a sync committee is assigned to slot `N`, they vote on what they see the block_root of slot `N` as. But this is then *included* in the block at slot `N+1`. So they vote on the current/assigned slot but when this makes it on chain it is a vote on the previous slot (same as an optimal attestation)

Reviewer: So I am looking at the object now and it seems that `slot` and the slot of `block` that is being signed over should be the same, whereas the comments imply otherwise. Essentially, at start of slot `block` is broadcast. Halfway through slot, sync committee votes. Next slot (`slot+1`) next block producer includes any sync committee votes from previous slot

Author: right, the comments are confusing -- i will update

Reviewer: Yeah, let's do that. It's how we assign `slot` to `Attestation` so fits our mental model already

    signature: BLSSignature
```

## Validator assignments

> A validator determines attestation committee assignments and beacon block proposal duties as defined in the Phase 0 document.

Reviewer: ```suggestion
A validator determines attestation committee assignments and beacon block proposal duties as defined in the Phase 0 document.
```

Author: did you have a suggestion here?

Reviewer: nope

> To determine sync committee assignments, a validator can run the following function: `get_sync_committee_assignment(state, epoch, validator_index)` where `epoch` is an epoch number within the current or next sync committee period. This function will return a tuple of `SyncCommitteeIndex` for each position in the sync committee that the pubkey corresponding to `validator_index` appears. If the validator has no assignments in the corresponding sync committees, this tuple will be empty `()`. 

Reviewer: Could use a sub-header "### Sync Committee"

```suggestion
### Sync Committee

To determine sync committee assignments, a validator can run the following function: `get_sync_committee_assignment(state, epoch, validator_index)` where `epoch` is an epoch number within the current or next sync committee period. This function will return a tuple of `SyncCommitteeIndex` for each position in the sync committee that the pubkey corresponding to `validator_index` appears. If the validator has no assignments in the corresponding sync committees, this tuple will be empty `()`. 
```

*Note*: `SyncCommitteeIndex` is different than the index into the validator registry or some attestation committee.

*Note*:  A single validator may be duplicated in a given sync committee.

```python
def compute_sync_committee_period(epoch: Epoch) -> uint64:
    return epoch // EPOCHS_PER_SYNC_COMMITTEE_PERIOD

def indices_in_sync_committee(sync_committee: SyncCommittee, target_pubkey: BLSPubkey) -> Tuple[SyncCommitteeIndex, ...]:
    return tuple(SyncCommitteeIndex(index) for index, pubkey in enumerate(sync_committee.pubkeys) if pubkey == target_pubkey)

> def get_sync_committee_assignment(state: BeaconState,

Reviewer: If we remove `SyncCommitteeIndex` and instead put this on the aggregator to include duplicates, this can just return a boolean. `return pubkey in state.next_sync_committee`

                                  epoch: Epoch,
                                  validator_index: ValidatorIndex) -> Tuple[SyncCommitteeIndex, ...]:
    sync_committee_period = compute_sync_committee_period(epoch)
    current_epoch = get_current_epoch(state)
    current_sync_committee_period = compute_sync_committee_period(current_epoch)
    next_sync_committee_period = current_sync_committee_period + 1
    assert sync_committee_period in (current_sync_committee_period, next_sync_committee_period)

    pubkey = state.validators[validator_index].pubkey
    if sync_committee_period == current_sync_committee_period:
        return indices_in_sync_committee(state.current_sync_committee, pubkey)
    else:  # sync_committee_period == next_sync_committee_period

>         return indices_in_sync_committee(state.next_sync_committee pubkey)

Reviewer: ```suggestion
        return indices_in_sync_committee(state.next_sync_committee, pubkey)
```

### Block proposal

Refer to the phase 0 document for the majority of the [block proposal responsibility](../phase0/validator.md#block-proposal). The validator should follow those instructions to prepare a `SignedBeaconBlock` for inclusion into the chain. All changes are additive to phase 0 and noted below.

#### Preparing a `BeaconBlock`

No change to [Preparing for a `BeaconBlock`](../phase0/validator.md#preparing-for-a-beaconblock).

#### Constructing the `BeaconBlockBody`

Each section of [Constructing the `BeaconBlockBody`](../phase0/validator.md#constructing-the-beaconblockbody) should be followed. After constructing the `BeaconBlockBody` as per that section, the proposer gains an additional task to include the sync committee signatures:

##### Sync committee

> The proposer will receive a number of `SyncCommitteeContribution`s from validators in the sync committee who are able to broadcast their signature. The proposer will collect all of these contributions for aggregation when preparing a block. `block.body.sync_committee_bits` is a `Bitvector` where the `i`th bit is `True` if the corresponding validator in the sync committee has produced a valid signature. `block.body.sync_committee_signature` is the aggregate BLS signature combining all the valid contributions. Given a collection of `contributions` and the `BeaconBlock` under construction, the proposer should process them as follows:

Reviewer: ```suggestion
The proposer receives a number of `SyncCommitteeContribution`s from validators in the sync committee who are able to broadcast their signature. The proposer then collects all of these contributions for aggregation when preparing a block. `block.body.sync_committee_bits` is a `Bitvector` where the `i`th bit is `True` if the corresponding validator in the sync committee has produced a valid signature. `block.body.sync_committee_signature` is the aggregate BLS signature combining all the valid contributions. Given a collection of `contributions` and the `BeaconBlock` under construction, the proposer should process them as follows:
```

> The proposer will receive a number of `SyncCommitteeContribution`s from validators in the sync committee who are able to broadcast their signature. The proposer will collect all of these contributions for aggregation when preparing a block. `block.body.sync_committee_bits` is a `Bitvector` where the `i`th bit is `True` if the corresponding validator in the sync committee has produced a valid signature. `block.body.sync_committee_signature` is the aggregate BLS signature combining all the valid contributions. Given a collection of `contributions` and the `BeaconBlock` under construction, the proposer should process them as follows:

Reviewer: If we remove `SyncCommitteeIndex` from the unaggregated message on the wire, the proposer would need to call `indices_in_sync_committee` and do any (maybe duplicate) inclusions for message they have received

Author: this work has been moved to the aggregator in the current design

```python

> def process_sync_committee_contributions(state: BeaconState, block: BeaconBlock, contributions: Sequence[SyncCommitteeContribution]) -> None:

Reviewer: Line length

Reviewer: Probably a `Set` instead of a `Sequence`

## Custom types

| Name | SSZ equivalent | Description |
| - | - | - |
| `SyncCommitteeIndex` | `uint64` | Index of a validator within a sync committee |

## Constants

### Misc

| Name | Value | Unit |
| - | - | :-: |
| `TARGET_AGGREGATORS_PER_SYNC_COMMITTEE` | `2**2` (= 4) | validators |
| `SYNC_COMMITTEE_SUBNET_COUNT` | `8` | The number of sync committee subnets used in the gossipsub protocol. |

## Containers

### SyncCommitteeSignature

```python
class SyncCommitteeSignature(Container):
    # Slot to which this contribution pertains
    slot: Slot
    # Index in the sync committee corresponding to the validator
    # who produced this signature
    index: SyncCommitteeIndex
    # Signature by the validator over the block root of `slot`
    signature: BLSSignature
```

### SyncCommitteeContribution

```python
class SyncCommitteeContribution(Container):
    # Slot to which this contribution pertains
    slot: Slot
    # A bit is set if a signature from the validator at the corresponding
    # index in the sync committee is present in the aggregate `signature`.

>     aggregate_bits: Bitvector[SYNC_COMMITTEE_SIZE]

Author: with the current scheme, only `1/SYNC_COMMITTEE_SUBNET_COUNT` of the bits will ever be set which at expected numbers is somewhat wasteful... we can further optimize this message so that there is no waste at the expense of an additional index (likely `uint64`) to indicate which "subcommittee" this message corresponds to.

i'm leaning towards omitting this as it would complicate the code further but open to hearing differing opinions on it

```python
def process_sync_committee_contributions(state: BeaconState, 
                                         block: BeaconBlock, 
                                         contributions: Set[SyncCommitteeContribution]) -> None:
    previous_slot = Slot(max(int(state.slot), 1) - 1)
    epoch = compute_epoch_at_slot(previous_slot)
    pubkeys = state.current_sync_committee.pubkeys
    domain = get_domain(state, DOMAIN_SYNC_COMMITTEE, epoch)
    signing_root = compute_signing_root(get_block_root_at_slot(state, previous_slot), domain)

    body = block.body
    signatures = []
    for contribution in contributions:
        participants = []
        for index, participated in enumerate(contribution.aggregate_bits):
            if participated:
                body.sync_committee_bits[index] = True
                participants.append(pubkeys[index])

        # is contribution valid?
        assert contribution.slot == previous_slot
        assert bls.FastAggregateVerify(participants, signing_root, contribution.signature)

        signatures.append(contribution.signature)

    body.sync_committee_signature = bls.Aggregate(signatures)
```

> *NOTE*: (TODO) likely to update with more context on subcommittee messages so we can skip some of the ~quadratic work here and save on "large" sparse messages

Author: see this: https://github.com/ethereum/eth2.0-specs/pull/2201/files#r583100236

HF 1 introduces a new type of committee: the sync committee. Sync committees are responsible for signing each block of the canonical chain and there exists an efficient algorithm for light clients to sync the chain using the output of the sync committees. See the [sync protocol](./sync-protocol.md) for further details on the light client sync. Under this network upgrade, validators track their participation in this new committee type and produce the relevant signatures as required. Block proposers incorporate the (aggregated) sync committee signatures into each block they produce.

## Prerequisites

All terminology, constants, functions, and protocol mechanics defined in the [HF 1 -- The Beacon Chain](./beacon-chain.md) doc are requisite for this document and used throughout. Please see this document before continuing and use as a reference throughout.

## Custom types

| Name | SSZ equivalent | Description |
| - | - | - |
| `SyncCommitteeIndex` | `uint64` | Index of a validator within a sync committee |

## Constants

### Misc

| Name | Value | Unit |
| - | - | :-: |

> | `TARGET_AGGREGATORS_PER_SYNC_COMMITTEE` | `2**2` (= 4) | validators |

Author: could argue this should be lower, say 2, but then you have very little redundancy in what is already a small subnet...

def get_contribution_and_proof(state: BeaconState,
                               aggregator_index: ValidatorIndex,
                               contribution: SyncCommitteeContribution,
                               privkey: int) -> ContributionAndProof:
    selection_proof = get_sync_committee_slot_signature(
      state, 
      contribution.slot,
      privkey
    )
    return ContributionAndProof(
        aggregator_index=aggregator_index,
        contribution=contribution,
        selection_proof=selection_proof,
    )
```

Then `signed_contribution_and_proof = SignedContributionAndProof(message=contribution_and_proof, signature=signature)` is constructed and broadast. Where `signature` is obtained from:

```python
def get_contribution_and_proof_signature(
  state: BeaconState,
  contribution_and_proof: ContributionAndProof,
  privkey: int) -> BLSSignature:
    contribution = contribution_and_proof.contribution
    domain = get_domain(state, DOMAIN_CONTRIBUTION_AND_PROOF, compute_epoch_at_slot(contribution.slot))
    signing_root = compute_signing_root(contribution_and_proof, domain)
    return bls.Sign(privkey, signing_root)
```

## Sync committee subnet stability

> - TODO: detail stability strategy, main thing is sharpness of transition...

Author: currently thinking to spec the next sync committee to join (all at once) some epochs before the end of the current sync committee period

                                  privkey: int) -> Sequence[SyncCommitteeSignature]:
    sync_committee = state.current_sync_committee
    pubkey = state.validators[validator_index].pubkey
    sync_committee_indices = compute_indices_in_sync_committee(sync_committee, pubkey)

    epoch = get_current_epoch(state)
    domain = get_domain(state, DOMAIN_SYNC_COMMITTEE, epoch)
    signing_root = compute_signing_root(get_block_root_at_slot(state, state.slot), domain)
    signature = bls.Sign(privkey, signing_root)

    return [
        SyncCommitteeSignature(
            slot=state.slot,
            index=index,
            signature=signature,
        ) for index in sync_committee_indices
    ]
```

##### Broadcast sync committee signature

The validator broadcasts the assembled signature(s) to the assigned subnet(s), the `sync_committee_{subnet_id}` pubsub topic.

The `subnet_id` is derived from the position in the sync committee such that the sync committee is divided into "subcommittees". It can be computed via `compute_subnet_for_sync_committee`:

```python
def compute_subnet_for_sync_committee(sync_committee_index: SyncCommitteeIndex) -> uint64:

>     return uint64(sync_committee_index % SYNC_COMMITTEE_SUBNET_COUNT)

Reviewer 3: Will there be modulo bias towards low subnets? Or does it divide nicely?

Author: checking this out: https://research.kudelskisecurity.com/2020/07/28/the-definitive-guide-to-modulo-bias-and-how-to-avoid-it/, it seems that we are ok here as we are moving b/t powers of two (1024 % 8)

#### Sync committee contributions

Each slot some sync committee members are selected to aggregate the `SyncCommitteeSignature`s into a `SyncCommitteeContribution` which is broadcast on a global channel for inclusion into the next block.

##### Aggregation selection

A validator is selected to aggregate based on the computation in `is_sync_committee_aggregator` where `state` is a `BeaconState` with the relevant sync committees for the queried `slot` (i.e. `state.slot` is within the span covered by the current or next sync committee period), and `signature` is the BLS signature returned by `get_sync_committee_slot_signature`. The signature function has the same parameters for `state` and `slot`; the `privkey` is the BLS private key associated with the validator.

```python
def get_sync_committee_slot_signature(state: BeaconState, slot: Slot, privkey: int) -> BLSSignature:
    domain = get_domain(state, DOMAIN_SYNC_COMMITTEE_SELECTION_PROOF, compute_epoch_at_slot(slot))
    signing_root = compute_signing_root(slot, domain)
    return bls.Sign(privkey, signing_root)
```

```python
def is_sync_committee_aggregator(state: BeaconState, slot: Slot, signature: BLSSignature) -> bool:

>     modulo = max(1, SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT // TARGET_AGGREGATORS_PER_SYNC_SUBCOMMITTEE)

Reviewer 3: Just double checking: `max(1, 1024 // 8 // 4) = max(1, 32) = 32`, 5 bits, constants would have to change a lot for any bias with 64 bit seed. LGTM.

HF 1 introduces a new type of committee: the sync committee. Sync committees are responsible for signing each block of the canonical chain and there exists an efficient algorithm for light clients to sync the chain using the output of the sync committees. See the [sync protocol](./sync-protocol.md) for further details on the light client sync. Under this network upgrade, validators track their participation in this new committee type and produce the relevant signatures as required. Block proposers incorporate the (aggregated) sync committee signatures into each block they produce.

## Prerequisites

All terminology, constants, functions, and protocol mechanics defined in the [HF 1 -- The Beacon Chain](./beacon-chain.md) doc are requisite for this document and used throughout. Please see this document before continuing and use as a reference throughout.

## Custom types

| Name | SSZ equivalent | Description |
| - | - | - |
| `SyncCommitteeIndex` | `uint64` | Index of a validator within a sync committee |

## Constants

### Misc

| Name | Value | Unit |
| - | - | :-: |

> | `TARGET_AGGREGATORS_PER_SYNC_COMMITTEE` | `2**2` (= 4) | validators |

Author: agreed, this was pulled from the attestation aggregation spec and didn't think too much about the name. agree we can update it for clarity

| `SYNC_COMMITTEE_SUBNET_COUNT` | `8` | The number of sync committee subnets used in the gossipsub protocol. |

## Containers

### SyncCommitteeSignature

```python
class SyncCommitteeSignature(Container):
    # Slot to which this contribution pertains
    slot: Slot
    # Index in the sync committee corresponding to the validator
    # who produced this signature

>     index: SyncCommitteeIndex

Author: are we ok w/ `index` if it refers to `ValidatorIndex`? or the more verbose `validator_index`?

>     index: SyncCommitteeIndex

Author: seems more consistent w/ rest of spec to be verbose here -- will update

```python
def process_sync_committee_contributions(state: BeaconState, 
                                         block: BeaconBlock, 
                                         contributions: Set[SyncCommitteeContribution]) -> None:
    previous_slot = Slot(max(int(state.slot), 1) - 1)
    epoch = compute_epoch_at_slot(previous_slot)
    pubkeys = state.current_sync_committee.pubkeys
    domain = get_domain(state, DOMAIN_SYNC_COMMITTEE, epoch)
    signing_root = compute_signing_root(get_block_root_at_slot(state, previous_slot), domain)

    body = block.body
    signatures = []
    for contribution in contributions:
        participants = []
        for index, participated in enumerate(contribution.aggregate_bits):
            if participated:
                body.sync_committee_bits[index] = True
                participants.append(pubkeys[index])

        # is contribution valid?
        assert contribution.slot == previous_slot
        assert bls.FastAggregateVerify(participants, signing_root, contribution.signature)

        signatures.append(contribution.signature)

    body.sync_committee_signature = bls.Aggregate(signatures)
```

> *NOTE*: (TODO) update w/ `SignedContributionAndProof` pending which validations go at gossip layer and which at application layer

Author: great, i'll be sure to clean up before merge

*NOTE*: (TODO) likely to update with more context on subcommittee messages so we can skip some of the ~quadratic work here and save on "large" sparse messages

> *NOTE*: (TODO) somewhere, the proposer needs to join _some_ gossip channel to collect signatures... have the proposer join one epoch ahead of schedule (or how long?)

Author: working on scheme to make the global channel as light as possible, can revisit more drastic changes to sync committee infra if the load becomes unmanageable

#### Packaging into a `SignedBeaconBlock`

No change to [Packaging into a `SignedBeaconBlock`](../phase0/validator.md#packaging-into-a-signedbeaconblock).

### Attesting and attestation aggregation

Refer to the phase 0 document for the [attesting](../phase0/validator.md#attesting) and [attestation aggregation](../phase0/validator.md#attestation-aggregation) responsibilities. There is no change compared to the phase 0 document.

### Sync committees

Sync committee members employ an aggregation scheme to reduce load on the proposer who is expected to include the full output of the committee every slot. Sync committee members produce individual signatures on subnets (like the attestation subnets via `SyncCommitteeSignatures` which are then collected by aggregators sampled from the sync committee to produce a `SyncCommitteeContribution` which is gossiped to proposers. This process occurs each slot.

> #### Sync committee signatures

Reviewer: STOPPED REVIEW HERE

    selection_proof = get_sync_committee_slot_signature(
      state, 
      contribution.slot,
      privkey
    )
    return ContributionAndProof(
        aggregator_index=aggregator_index,
        contribution=contribution,
        selection_proof=selection_proof,
    )
```

Then `signed_contribution_and_proof = SignedContributionAndProof(message=contribution_and_proof, signature=signature)` is constructed and broadast. Where `signature` is obtained from:

```python
def get_contribution_and_proof_signature(
  state: BeaconState,
  contribution_and_proof: ContributionAndProof,
  privkey: int) -> BLSSignature:
    contribution = contribution_and_proof.contribution
    domain = get_domain(state, DOMAIN_CONTRIBUTION_AND_PROOF, compute_epoch_at_slot(contribution.slot))
    signing_root = compute_signing_root(contribution_and_proof, domain)
    return bls.Sign(privkey, signing_root)
```

## Sync committee subnet stability

The sync committee subnets need special care to ensure stability given the relatively low number of validators involved in the sync committee at any particular time. To provide this stability, a validator must do the following:

> * Maintain advertisement of the subnet the validator in the sync committee is assigned to in their node's ENR as soon as they have joined the subnet. Subnet assignments are known `EPOCHS_PER_SYNC_COMMITTEE_PERIOD` epochs in advance and can be computed with `compute_subnet_for_sync_committee` defined above. ENR advertisement is indicated by setting the `syncnets` entry in the ENR to a `uint8` encoding of the `subnet_id`, e.g. `ENR["syncnets"] = b'\x07'`. This entry is removed from the ENR after the validator has left the sync committee.

Author: a consideration here is that ENRs are very space limited but we also don't have a better way to advertise sync committee subscriptions at the moment...

Reviewer: we could extend `attnets` and just make it `subnets`, where first 64 are attnets and next 8 bits are sync committee subnets

> - TODO: what happens for the first sync committee?

Author: may need to advertise as soon as you know which subnet you are in (so 1 sync committee period of lookahead) in order to give validating nodes time to find peers... seems like you can still join just ahead of the committee change but i'd have to revisit gossipsub to make sure this is feasible

Reviewer: hmm, probably want to join ~1 to 5 epochs head of time and advertise when you join.

> * To join a sync committee subnet, select a random number of epochs before the end of the current sync committee period between 1 and `SYNC_COMMITTEE_SUBNET_COUNT`, inclusive. Validators should join their member subnet at the beginning of the epoch they have randomly selected. For example, if the next sync committee period starts at epoch `853,248` and the validator randomly selects an offset of `3`, they should join the subnet at the beginning of epoch `853,245`.

Author: the thinking here is to smooth out the "sharpness" of the transition of having one chunk of the sync committee suddenly join a subnet. an alternative that may actually be fine (and is simpler to implement) is to just have the entire next sync committee join their respective subnet at a fixed epoch ahead of time, e.g. `boundary_epoch` - 8 epochs

## Prerequisites

All terminology, constants, functions, and protocol mechanics defined in the [HF1 -- The Beacon Chain](./beacon-chain.md) doc are requisite for this document and used throughout. Please see this document before continuing and use as a reference throughout.

## Custom types

| Name | SSZ equivalent | Description |
| - | - | - |
| `SyncCommitteeIndex` | `uint64` | Index of a validator within a sync committee |

## Constants

### Misc

| Name | Value | Unit |
| - | - | :-: |
| `TARGET_AGGREGATORS_PER_SYNC_COMMITTEE` | `2**2` (= 4) | validators |
| `SYNC_COMMITTEE_SUBNET_COUNT` | `8` | The number of sync committee subnets used in the gossipsub aggregation protocol. |

## Containers

### SyncCommitteeSignature

```python
class SyncCommitteeSignature(Container):
    # Slot to which this contribution pertains
    slot: Slot
    # Index in the sync committee corresponding to the validator that produced this signature
    index: SyncCommitteeIndex
    # Signature by the validator over the block root of `slot`
    signature: BLSSignature
```

### SyncCommitteeContribution

```python
class SyncCommitteeContribution(Container):
    # Slot to which this contribution pertains

>     slot: Slot

Reviewer: Similarly, I think we'll want block_root here. but we can debate a bit more tomorrow

            index=index,
            signature=signature,
        ) for index in sync_committee_indices
    ]
```

##### Broadcast sync committee signature

The validator broadcasts the assembled signature(s) to the assigned subnet(s), the `sync_committee_{subnet_id}` pubsub topic.

The `subnet_id` is derived from the position in the sync committee such that the sync committee is divided into "subcommittees". It can be computed via `compute_subnet_for_sync_committee`:

```python
def compute_subnet_for_sync_committee(sync_committee_index: SyncCommitteeIndex) -> uint64:
    return uint64(sync_committee_index % SYNC_COMMITTEE_SUBNET_COUNT)
```

*NOTE*: subnet assignment does not change during the duration of a validator's assignment to a given sync committee.

*NOTE*: (TODO) revisit subnet algo -- may be a bit more efficient to make "contiguous" "subcommittees"
*NOTE*: (TODO) details of when to join subnets

#### Sync committee contributions

Each slot some sync committee members are selected to aggregate the `SyncCommitteeSignature`s into a `SyncCommitteeContribution` which is broadcast on a global channel for inclusion into the next block.

##### Aggregation selection

> A validator is selected to aggregate based on the computation in `is_sync_committee_aggregator` where `state` is a `BeaconState` with the relevant sync committees for the queried `slot`, i.e. `state.slot` is within the span covered by the current or next sync committee period. `privkey` is the BLS private key associated with the validator.

Reviewer: ```suggestion
A validator is selected to aggregate based on the computation in `is_sync_committee_aggregator` where `state` is a `BeaconState` with the relevant sync committees for the queried `slot`, i.e. `state.slot`, is within the span covered by the current or next sync committee period. `privkey` is the BLS private key associated with the validator.
```

```python
def get_sync_committee_slot_signature(state: BeaconState, slot: Slot, privkey: int) -> BLSSignature:

>     domain = get_domain(

Reviewer: prefer one liner when possible (only 98 chars of the 120 limit)

      state, 
      DOMAIN_SYNC_COMMITTEE_SELECTION_PROOF, 
      compute_epoch_at_slot(slot)
    )
    signing_root = compute_signing_root(slot, domain)
    return bls.Sign(privkey, signing_root)

> 

Reviewer: separate code blocks (same formatting as the code blocks in [attestation aggregation selection](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/validator.md#aggregation-selection)

def is_sync_committee_aggregator(state: BeaconState, slot: Slot, privkey: int) -> bool:
    signature = get_sync_committee_slot_signature(state, slot, privkey)

>     modulo = max(

Reviewer: I think one liner

      1, 

>       SYNC_COMMITTEE_SIZE // TARGET_AGGREGATORS_PER_SYNC_COMMITTEE

Reviewer: Ah this looks way too low, right? We want to target 2 or 4 aggregators per *sub committee* rather than just 4 for the entire committee.

Should we make this function be for the sub-committee? and the target value based on the sub committee size?

Author: yep, this should be updated... let me fix

    )
    return bytes_to_uint64(hash(signature)[0:8]) % modulo == 0
```

> *NOTE*: the set of aggregators will generally change every slot; however, the assignments can be computed ahead of time as soon as the committee is known

Reviewer: ```suggestion
*NOTE*: the set of aggregators generally changes every slot; however, the assignments can be computed ahead of time as soon as the committee is known
```

##### Construct sync committee contribution

If a validator is selected to aggregate the `SyncCommitteeSignature`s produced on a subnet during a given `slot`, they construct an aggregated `SyncCommitteeContribution`. Given all of the collected `sync_committee_signatures` from the `sync_committee_{subnet_id}` gossip, create the contribution via `construct_sync_committee_contribution`. `state` in this function is the same `BeaconState` used by the aggregator to produce their individual signature for the assigned slot.

> 

Reviewer: Slight preference for more granular setting of fields as was done in [validator guide aggregation](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/validator.md#construct-aggregate)

not a big deal either way but a bit easier to parse I think

```python

> def construct_sync_committee_contribution(state: BeaconState, sync_committee_signatures: Set[SyncCommitteeSignature]) -> SyncCommitteeContribution:

Reviewer: line length

    contribution = SyncCommitteeContribution()
    contribution.slot = state.slot

    sync_committee = state.current_sync_committee

>     epoch = compute_epoch_at_slot(state.slot)

Reviewer: use `get_current_epoch`

    domain = get_domain(state, DOMAIN_SYNC_COMMITTEE, epoch)

>     signing_root = compute_signing_root(

Reviewer: one liner

      get_block_root_at_slot(state, state.slot), 
      domain
    )

    signatures = []

>     for sync_committee_signature in sync_committee_signatures:

Reviewer: If we use block_root, the aggregator will need to filter the committee signatures by the block_root of their own contribution. This doesn't have to be in code. Can jsut be a note about the sync_committee_signatures that are supposed to be passed in like [here](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/validator.md#construct-aggregate)

        index = sync_committee_signature.index
        signature = sync_committee_signature.signature
        contribution.aggregate_bits[index] = True
        signatures.append(signature)

>         # is sync committee signature valid?

Reviewer: spec code comments use capital letter to start

```suggestion
        # Is sync committee signature valid?
```

        assert sync_committee_signature.slot == state.slot

>         assert bls.Verify(

Reviewer: I don't think we need assertions at this point in the spec. The conditions that these values would be passed into this function would be that they were verified during gossip validations. I would just note that you collected the valid `SyncCommitteeSignature`s for the slot and leave the validations out

Author: will remove as i add gossip conditions

Selection proofs are provided in `ContributionAndProof` to prove to the gossip channel that the validator has been selected as an aggregator.

`ContributionAndProof` messages are signed by the aggregator and broadcast inside of `SignedContributionAndProof` objects to prevent a class of DoS attacks and message forgeries.

First, `contribution_and_proof = get_contribution_and_proof(state, validator_index, contribution, privkey)` is constructed.

```python
def get_contribution_and_proof(state: BeaconState,
                               aggregator_index: ValidatorIndex,
                               contribution: SyncCommitteeContribution,
                               privkey: int) -> ContributionAndProof:
    selection_proof = get_sync_committee_slot_signature(
      state, 
      contribution.slot,
      privkey
    )
    return ContributionAndProof(
        aggregator_index=aggregator_index,
        contribution=contribution,

>         selection_proof=selection_proof,

Reviewer: inline preference
```suggestion
        selection_proof=get_sync_committee_slot_signature(state, contribution.slot, privkey),
```

    )
```

Then `signed_contribution_and_proof = SignedContributionAndProof(message=contribution_and_proof, signature=signature)` is constructed and broadast. Where `signature` is obtained from:

```python
def get_contribution_and_proof_signature(

>   state: BeaconState,

Reviewer: spacing. prefer 4-spaces and prefer trailing off the function name like above (`get_contribution_and_proof`)

  contribution_and_proof: ContributionAndProof,
  privkey: int) -> BLSSignature:
    contribution = contribution_and_proof.contribution
    domain = get_domain(state, DOMAIN_CONTRIBUTION_AND_PROOF, compute_epoch_at_slot(contribution.slot))
    signing_root = compute_signing_root(contribution_and_proof, domain)
    return bls.Sign(privkey, signing_root)
```

## Sync committee subnet stability

The sync committee subnets need special care to ensure stability given the relatively low number of validators involved in the sync committee at any particular time. To provide this stability, a validator must do the following:

> * Maintain advertisement of the subnet the validator in the sync committee is assigned to in their node's ENR as soon as they have joined the subnet. Subnet assignments are known `EPOCHS_PER_SYNC_COMMITTEE_PERIOD` epochs in advance and can be computed with `compute_subnet_for_sync_committee` defined above. ENR advertisement is indicated by setting the `syncnets` entry in the ENR to a `uint8` encoding of the `subnet_id`, e.g. `ENR["syncnets"] = b'\x07'`. This entry is removed from the ENR after the validator has left the sync committee.

Reviewer: damn, privacy on this shit is terrible

> * Maintain advertisement of the subnet the validator in the sync committee is assigned to in their node's ENR as soon as they have joined the subnet. Subnet assignments are known `EPOCHS_PER_SYNC_COMMITTEE_PERIOD` epochs in advance and can be computed with `compute_subnet_for_sync_committee` defined above. ENR advertisement is indicated by setting the `syncnets` entry in the ENR to a `uint8` encoding of the `subnet_id`, e.g. `ENR["syncnets"] = b'\x07'`. This entry is removed from the ENR after the validator has left the sync committee.

Reviewer: This should probably be a bitfield. There's a chance you are only multiple subnets at once due to duplicate assignments or running multilpe validators on your machine

Author: yes, good point
